lib/spack/spack/spec.py

```diff
@@ -750,6 +753,162 @@ def __str__(self):
         return "{deps: %s}" % ', '.join(str(d) for d in sorted(self.values()))
 
 
+def _libs_default_handler(descriptor, spec, cls):
+    """Default handler when looking for 'libs' attribute. The default
+    tries to search for 'lib{spec.name}' recursively starting from
+    `spec.prefix`.
+
+    :param ForwardQueryToPackage descriptor: descriptor that triggered
+        the call
+    :param Spec spec: spec that is being queried
+    :param type(spec) cls: type of spec, to match the signature of the
+        descriptor `__get__` method
+    """
+    name = 'lib' + spec.name
+    shared = '+shared' in spec
+    return find_libraries(
+        [name], root=spec.prefix, shared=shared, recurse=True
+    )
+
+
+def _cppflags_default_handler(descriptor, spec, cls):
+    """Default handler when looking for cppflags attribute. The default
+    just returns '-I{spec.prefix.include}'.
+
+    :param ForwardQueryToPackage descriptor: descriptor that triggered
+        the call
+    :param Spec spec: spec that is being queried
+    :param type(spec) cls: type of spec, to match the signature of the
+        descriptor `__get__` method
+    """
+    return '-I' + spec.prefix.include
+
+
+class ForwardQueryToPackage(object):
+    """Descriptor used to forward queries from Spec to Package"""
+
+    def __init__(self, attribute_name, default_handler=None):
+        """Initializes the instance of the descriptor
+
+        :param str attribute_name: name of the attribute to be
+            searched for in the Package instance
+        :param callable default_handler: [optional] default function
+            to be called if the attribute was not found in the Package
+            instance
+        """
+        self.attribute_name = attribute_name
+        # Turn the default handler into a function with the right
+        # signature that always returns None
+        if default_handler is None:
+            default_handler = lambda descriptor, spec, cls: None
+        self.default = default_handler
+
+    def __get__(self, instance, cls):
+        """Retrieves the property from Package using a well defined chain
+        of responsibility.
+
+        The order of call is :
+
+        1. if the query was through the name of a virtual package try to
+            search for the attribute `{virtual_name}_{attribute_name}`
+            in Package
+
+        2. try to search for attribute `{attribute_name}` in Package
+
+        3. try to call the default handler
+
+        The first call that produces a value that is not None will stop the
+        chain.
+
+        If no call can handle the request or a None value is produced,
+        then AttributeError is raised
+        """
```

Comment:
@adamjstewart ... and in the docstring above you have a description of the querying mechanism.